pkg/util/strategicpatch/patch.go

```diff
@@ -31,7 +31,352 @@ import (
 // lists should be merged or replaced.
 //
 // For more information, see the PATCH section of docs/api-conventions.md.
+
+// CreateStrategicMergePatch creates a patch that can be passed to StrategicMergePatch.
```

Comment:
Just had a long chat with @jackgr. It looks like we may be on the wrong track with this approach. Based on what is currently implemented in this PR, here is how `kubectl apply` would work:

1. Get the live object from the server along with the original annotation.
1. Compare the local object to the original annotation and *create a patch* of the differences between the two. Then send that patch to the server as a strategic merge patch.

The issue with this approach is the use case for when I run `kubectl apply` on a resource that I have not changed since the last configuration run, but the live server state has modified fields that are explicitly defined in my local object. This should not be a conflict - any and all fields specified in our local object should *always* simply override all the server object fields. The above approach does not accomplish this because it has no way to detect such changes to the server object. Therefore it does not give us true consistent reconciliation of the local config object to the server every time we run `kubectl apply`.

The alternative approach (originally outlined in https://github.com/kubernetes/kubernetes/issues/1702#issuecomment-89743530) is as follows:

1. Get the live object from the server along with the original annotation.
1. Compare the local object with the original annotation to generate a new JSON object which is equivalent to the current local object but decorated with null's (or `$patch: delete`'s) needed to remove any fields that were removed in the current local object but existed in the original annotation.
1. Submit that new JSON as a Strategic Merge patch.

In other words, the local config object IS the patch itself sans deletions of removed fields. So the only thing we need to add is a method like AddDeletionsToPatch, which would take the local object and the original annotation as parameters and return the local object decorated with the right deletions (e.g. mapKey: null and $patch: delete).

@bgrant0607 can you confirm that the second approach is what we actually want and we should go in that direction instead?